src/compiler/gcc.rs

```diff
@@ -512,8 +530,11 @@ pub fn generate_compile_commands(
             "-o".into(),
             path_transformer.to_dist(out_file)?,
         ];
-        // We could do preprocessor_args here, but skip for consistency with msvc
         arguments.extend(dist::osstrings_to_strings(&parsed_args.common_args)?);
+        // We need to include the preprocessor_args here, since the preprocessor run is
+        // configured to only resolve includes, so the preprocessor needs to run on the
+        // dist side again.
```

Comment:
Would it be possible to do a quick benchmark of the impact of this? It may be worth it even if it adds some overhead, but it would be good to know.

Reply:
I am not sure how to do a meaningful benchmark. It would need to use a containerized scheduler & server to be somewhat controllable. It would certainly be interesting to know, so with some guidance on how to perform such a benchmark, I can surely do that!

However, IMHO the modifications by this PR are essential for regular use cases, both locally and in CI. Distributing a build is usually expected to be transparent, and without these changes, transparency is very much broken: Any error and warning messages involving macros become hardly comprehensible, and it is essentially impossible to run a build with -Werror.

Reply:
I tend to agree with @sigiesec. I was going to take some measurements by doing mozilla-central clobber builds, but it seems this needs more work as this breaks mozilla-central builds completely...

In particular, mozilla-central uses `-MD -MP -MF file.pp`, and with this patch those flags are passed to the remote builder (and it isn't there of course).

So I think rather than this, we should just make the `-D` flag a `PassThrough` flag rather than a `PreprocessorArg` so it goes to common_args and gets passed both to the preprocessor and the compiler, but without passing the other preprocessor flags like `-i*` and `-M*` flags.

Reply:
So we should probably add another test case (or add it to the existing test cases) using these options, so we catch these issues in the sccache tests already :)

Reply:
Hmm, doing that I get further along, but my build fails in a somewhat mysterious way in a system header...

Reply:
So the command that's failing is `/home/emilio/.cargo/bin/sccache /home/emilio/.mozbuild/clang/bin/clang++ -o nsLocalFileUnix.o -c  -I/home/emilio/src/moz/gecko-2/obj-debug/dist/stl_wrappers -I/home/emilio/src/moz/gecko-2/obj-debug/dist/system_wrappers -include /home/emilio/src/moz/gecko-2/config/gcc_hidden.h -DDEBUG=1 -DOS_POSIX=1 -DOS_LINUX=1 -DSTATIC_EXPORTABLE_JS_API -DMOZ_HAS_MOZGLUE -DMOZILLA_INTERNAL_API -DIMPL_LIBXUL -I/home/emilio/src/moz/gecko-2/xpcom/io -I/home/emilio/src/moz/gecko-2/obj-debug/xpcom/io -I/home/emilio/src/moz/gecko-2/obj-debug/ipc/ipdl/_ipdlheaders -I/home/emilio/src/moz/gecko-2/ipc/chromium/src -I/home/emilio/src/moz/gecko-2/ipc/glue -I/home/emilio/src/moz/gecko-2/obj-debug/xpcom -I/home/emilio/src/moz/gecko-2/xpcom/build -I/home/emilio/src/moz/gecko-2/obj-debug/dist/include -I/home/emilio/src/moz/gecko-2/obj-debug/dist/include/nspr -I/home/emilio/src/moz/gecko-2/obj-debug/dist/include/nss -fPIC -DMOZILLA_CLIENT -include /home/emilio/src/moz/gecko-2/obj-debug/mozilla-config.h -Qunused-arguments -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -fstack-protector-strong -ftrivial-auto-var-init=pattern -Qunused-arguments -Wall -Wbitfield-enum-conversion -Wempty-body -Wignored-qualifiers -Woverloaded-virtual -Wpointer-arith -Wshadow-field-in-constructor-modified -Wsign-compare -Wtype-limits -Wunreachable-code -Wunreachable-code-return -Wwrite-strings -Wno-invalid-offsetof -Wclass-varargs -Wfloat-overflow-conversion -Wfloat-zero-conversion -Wloop-analysis -Wc++1z-compat -Wc++2a-compat -Wcomma -Wimplicit-fallthrough -Wstring-conversion -Wtautological-overlap-compare -Wtautological-unsigned-enum-zero-compare -Wtautological-unsigned-zero-compare -Wno-error=tautological-type-limit-compare -Wno-inline-new-delete -Wno-error=deprecated-declarations -Wno-error=array-bounds -Wno-error=backend-plugin -Wno-error=return-std-move -Wno-error=atomic-alignment -Wformat -Wformat-security -Wno-gnu-zero-variadic-macro-arguments -Wno-unknown-warning-option -Wno-return-type-c-linkage -fno-sized-deallocation -fno-aligned-new -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -fstack-protector-strong -ftrivial-auto-var-init=pattern -fno-exceptions -fno-strict-aliasing -fno-rtti -fno-exceptions -fno-math-errno -pthread -pipe -g -Os -fno-omit-frame-pointer -funwind-tables  -MD -MP -MF .deps/nsLocalFileUnix.o.pp  -fcolor-diagnostics  /home/emilio/src/moz/gecko-2/xpcom/io/nsLocalFileUnix.cpp`

And it does indeed compile if I remove the sccache wrapper. Not sure yet why.

Reply:
Yeah I'm trying to reduce the issue with that suggestion too.

Reply:
So the reduced test-case is a pile of sadness like this:

```c
 #define a __has_include                                                                                                                                                                                                                     
extern                                                                           
  #if a("linux/stat.h")                                                            
      ;                                                                            
  #endif
```

I think this is a bug in `clang`'s `-frewrite-include`. The tldr is that my glibc has something like:

```cpp
#ifdef __has_include
#define __glibc_has_include(header) __has_include(header)
#else 
#define __glibc_has_include(header) 0
#endif 

#if __glibc_has_include("linux/stat.h")
#include <linux/stat.h>
#endif
```

And then there's some code in gecko that includes that file unconditionally.

But clang rewrites the first include and not the second, but _doesn't_ remove the `__glibc_has_include` condition, so the builder, which doesn't have that header, doesn't process the included header and everything falls apart... GCC works fine, but GCC + sccache give me a `fatal error: cannot execute 'as': execvp: No such file or directory`.

Reply:
So I think we should change this patch to do something like:

```diff
diff --git a/src/compiler/gcc.rs b/src/compiler/gcc.rs
index 8ef663b..e7a1354 100644
--- a/src/compiler/gcc.rs
+++ b/src/compiler/gcc.rs
@@ -121,7 +121,7 @@ counted_array!(pub static ARGS: [ArgInfo<ArgData>; _] = [
     take_arg!("--sysroot", PathBuf, Separated, PassThroughPath),
     take_arg!("-A", OsString, Separated, PassThrough),
     take_arg!("-B", PathBuf, CanBeSeparated, PassThroughPath),
-    take_arg!("-D", OsString, CanBeSeparated, PreprocessorArgument),
+    take_arg!("-D", OsString, CanBeSeparated, PassThrough),
     flag!("-E", TooHardFlag),
     take_arg!("-F", PathBuf, CanBeSeparated, PreprocessorArgumentPath),
     take_arg!("-G", OsString, Separated, PassThrough),
@@ -136,7 +136,7 @@ counted_array!(pub static ARGS: [ArgInfo<ArgData>; _] = [
     take_arg!("-MQ", OsString, Separated, PreprocessorArgument),
     take_arg!("-MT", OsString, Separated, DepTarget),
     flag!("-P", TooHardFlag),
-    take_arg!("-U", OsString, CanBeSeparated, PreprocessorArgument),
+    take_arg!("-U", OsString, CanBeSeparated, PassThrough),
     take_arg!("-V", OsString, Separated, PassThrough),
     take_arg!("-Xassembler", OsString, Separated, PassThrough),
     take_arg!("-Xlinker", OsString, Separated, PassThrough),
@@ -549,10 +549,6 @@ pub fn generate_compile_commands(
             path_transformer.to_dist(out_file)?,
         ];
         arguments.extend(dist::osstrings_to_strings(&parsed_args.common_args)?);
-        // We need to include the preprocessor_args here, since the preprocessor run is
-        // configured to only resolve includes, so the preprocessor needs to run on the
-        // dist side again.
-        arguments.extend(dist::osstrings_to_strings(&parsed_args.preprocessor_args)?);
         Some(dist::CompileCommand {
             executable: path_transformer.to_dist(&executable)?,
             arguments: arguments,
```

And then we should go:

 * Fix g++ so that it works right with sccache-dist.
 * Fix clang++'s `-frewrite-includes` to handle the above situation correctly.

Reply:
So to elaborate, a more comprehensive test-case would be:

```cpp
#define has_include(header) __has_include(header)                                                                                                                                                                                           
                                                                                  
#if has_include("t.h")                                                           
#include <iostream>                                                              
#endif 

#include <iostream>                                                              
                                                                                   
int main() {                                                                     
  std::cout << "Hello world" << std::endl;                                       
  return 0;                                                                      
}
```

Where `t.h` is an empty file locally.

`clang -frewrite-includes` will leave the `has_include` check, but it will _not_ expand the second `#include` (this behavior seems to be limited to system headers somehow).

Reply:
So I filed a clang bug in:

 * https://bugs.llvm.org/show_bug.cgi?id=43982

But right after doing so I realized of a similar bug:

 * https://bugs.llvm.org/show_bug.cgi?id=37990

Where Richard Smith explains that this is invalid per http://eel.is/c++draft/cpp.cond#7.sentence-2.

So I ended up filing a libc bug:

 * https://sourceware.org/bugzilla/show_bug.cgi?id=25189

And sending a patch because why not:

 * https://sourceware.org/ml/libc-alpha/2019-11/msg00431.html

Reply:
Ok, with that glibc patch, and the diff I posted above applied to your tree, I can build Firefox without issues \o/

Reply:
And I took some numbers, and these patches + mine make the build considerably faster (maybe due to the reduced output from the compiler).

I got a clobber debug build in Berlin in:

> 4:30.65 Your build was successful!

:)

Reply:
Sadly, you can't rely on people having a new enough glibc once the patch lands.

Reply:
> And I took some numbers, and these patches + mine make the build considerably faster (maybe due to the reduced output from the compiler).
> 
> I got a clobber debug build in Berlin in:
> 
> > 4:30.65 Your build was successful!
> 
> :)

That's impressive, was the cache empty for that build?

Reply:
> I am not sure how to do a meaningful benchmark. It would need to use a containerized scheduler & server to be somewhat controllable. It would certainly be interesting to know, so with some guidance on how to perform such a benchmark, I can surely do that!
> 
> However, IMHO the modifications by this PR are essential for regular use cases, both locally and in CI. Distributing a build is usually expected to be transparent, and without these changes, transparency is very much broken: Any error and warning messages involving macros become hardly comprehensible, and it is essentially impossible to run a build with -Werror.

Yes, I agree, I just meant to try running a build with and without to see if it adds considerable overhead. But looking at this patch again I don't think it's going to result in too much extra work.

Reply:
> Sadly, you can't rely on people having a new enough glibc once the patch lands.

I think the glibc version on my machine is pre-release. So assuming that's the case (not 100% sure) I think it's ok.

Reply:
Gah, I lie, the problematic code is is in glibc 2.30 :(

Reply:
We can certainly not land the patch relying on some very-recent/future glibc. What confuses me here is that this has been working with icecream (which uses `-frewrite-includes`/`-fdirectives-only`, so I think we are still missing something.

Hearing that this actually speeds up compilation is great! This might also explain why I had the impression that sccache has been slower than icecream despite the support for rust.

Reply:
FWIW, these commits implemented this for icecream, for clang resp. gcc:
* clang: https://github.com/icecc/icecream/commit/b78ca31a18cbd24efe5a015da4734dc72afeb374 (note that the clang option was `-Wp,-rewrite-includes` originally, but was later changed to `-frewrite-includes`)
* gcc: https://github.com/icecc/icecream/commit/7ab429b5f3405ec8451c31f80573658c8818b972

This is on by default, but suppressed when -Wunused-macros is used, via https://github.com/icecc/icecream/commit/4f46699b945ed26fb9924609fdf68e606b0b53b7

There is also code to force local recompilation when there was an error remotely with gcc only due to some limitations of `-fdirectives-only`, added via https://github.com/icecc/icecream/commit/861030d617086560dacce2f000df5db0f7f19b9a

But I assume that @emilio's problem with glibc happened with clang?

Reply:
Yes, my issue happened with clang. But it's a pretty recent issue, I suspect it'd happen with icecc as well.

Reply:
By "recent" you mean the problematic `__has_include` usage was only introduced in a recent version of glibc?

Reply:
Can we put this behind an option, as icecc did, and document the issue with glibc?

Reply:
@chmanchester Makes sense. I will add that. Maybe at sometime in the future we can default it to on then.

Reply:
So this means a (boolean) field needs to be added to DistConfig, right? Does this automagically generate parser code for the config file? I didn't find code that parser it explicitly.

Reply:
> So this means a (boolean) field needs to be added to DistConfig, right? Does this automagically generate parser code for the config file? I didn't find code that parser it explicitly.

It's using serde to `derive(Serialize, Deserialize)` so you get that for free:
https://github.com/mozilla/sccache/blob/a2cab6b35fd9cec47d09275a2448a272d771c82c/src/config.rs#L374